odoo/addons/base/models/res_users.py

```diff
@@ -2306,6 +2317,19 @@ def init(self):
             table,
         ))
 
+    def _inverse_expiration_date(self):
+        max_duration = max(
+            datetime.timedelta(**{group.apikey_duration_unit: group.apikey_duration_value})
+            for group in self.env.user.groups_id
+        )
+        max_expiration_date = datetime.datetime.now() + max_duration
+        for apikey in self:
+            if not apikey.expiration_date and not self.env.is_system():
+                raise UserError(_("Only an Administrator can remove the expiration date."))
+            elif apikey.expiration_date and apikey.expiration_date > max_expiration_date:
+                raise UserError(_("You cannot exceed %(days)d days from today for API key %(name)s.",
+                    days=max_duration.days, name=apikey.name))
```

Comment:
This approach is more secure.
However, the frontend needs to be modified so that users can add an expiry date (depending on their privileges).
There would no longer be any need to manage API key updates (so no need to add access rights).

If we opt for this solution, we will need to design the logic to recreate a new key (an email or something else).
But this would have the advantage of rechecking the user's privileges (each time the key expires).

@beledouxdenis What do you think ?

Reply:
After internal discussion, we will opt for a fixed expiry date that is determined when the API key is created.